src/interval.rs

```diff
@@ -0,0 +1,785 @@
+            left,
+            right,
+            left_infinite: false,
+            right_infinite: false,
+        }
+    }
+
+    /// Computes the length of the interval. Returns none if any of the bounds
+    /// are infinite.
+    #[allow(dead_code)]
+    pub fn length(&self) -> Option<T> {
+        if self.left_infinite || self.right_infinite {
+            None
+        } else {
+            let (min, max) = if self.right > self.left {
+                (self.left, self.right)
+            } else {
+                (self.right, self.left)
+            };
+            Some(max - min)
+        }
+    }
+
+    /// Returns the left limit of the interval or None if left is infinite.
+    #[allow(dead_code)]
+    pub fn left(&self) -> Option<T> {
+        if self.left_infinite {
+            None
+        } else {
+            Some(self.left)
+        }
+    }
+
+    /// Returns the right limit of the interval or None if right is infinite.
+    #[allow(dead_code)]
+    pub fn right(&self) -> Option<T> {
+        if self.right_infinite {
+            None
+        } else {
+            Some(self.right)
+        }
+    }
+
+    /// Computes the length of the interval. Returns none if any of the bounds
+    /// are infinite.
+    pub fn centre(&self) -> Option<T> {
+        if self.left_infinite || self.right_infinite {
+            None
+        } else {
+            let (min, max) = if self.right > self.left {
+                (self.left, self.right)
+            } else {
+                (self.right, self.left)
+            };
+            Some((min + max) / T::from_f32(2_f32).unwrap())
+        }
+    }
+
+    /// Checks if the open interval includes the input value. If the value
+    /// equals any of the extremities, it is not considered included.
+    #[allow(dead_code)]
+    pub fn includes_open(&self, value: T) -> bool {
+        if self.left_infinite && self.right_infinite {
+            return true;
+        }
+        if self.left_infinite {
+            return value < self.right;
+        }
+        if self.right_infinite {
+            return value > self.left;
+        }
+        let (min, max) = if self.right > self.left {
+            (self.left, self.right)
+        } else {
+            (self.right, self.left)
+        };
+        value > min && value < max
+    }
+
+    /// Checks if the closed interval includes the input value. If the value
+    /// equals any of the extremities, it is considered included.
+    pub fn includes_closed(&self, value: T) -> bool {
+        if self.left_infinite && self.right_infinite {
+            return true;
+        }
+        if self.left_infinite {
+            return value <= self.right;
+        }
+        if self.right_infinite {
+            return value >= self.left;
+        }
+        let (min, max) = if self.right > self.left {
+            (self.left, self.right)
+        } else {
+            (self.right, self.left)
+        };
+        value >= min && value <= max
+    }
```

Comment:
I was hesitating. I
I'm not sure how often will we read the same interval in various ways. Maybe never.
Regarding Interval vs. Range, this was my thinking (but I don't insist on keeping the Interval):

- Range allows for iterating over the values. This is semantically not possible with interval.
- Range doesn't allow for left-open intervals such as `=a..b` (or am I missing something?)
- We'd have to implement custom functions, such as `length`, `centre` and `remap`

Reply:
> Range allows for iterating over the values. This is semantically not possible with Interval.

I don't think it should trouble us if we don't use it. No need to prevent it.

> Range doesn't allow for left-open intervals such as =a..b (or am I missing something?)

Looks like there is no literal for that indeed. The `RangeBounds` trait does support it though, and this impl exists: `impl<T> RangeBounds<T> for (Bound<T>, Bound<T>)`

> We'd have to implement custom functions, such as length, centre and remap

From what I remember seeing in the PR, only `remap`. `centre` and `length` were just used to power remap. (I might remember incorrectly, or this fact could have changed since the last review).